components/salsa-macros/src/tracked_fn.rs

```diff
@@ -201,25 +203,38 @@ impl Macro {
     fn cycle_recovery(&self) -> syn::Result<(TokenStream, TokenStream, TokenStream)> {
         // TODO should we ask the user to specify a struct that impls a trait with two methods,
         // rather than asking for two methods separately?
-        match (&self.args.cycle_fn, &self.args.cycle_initial) {
-            (Some(cycle_fn), Some(cycle_initial)) => Ok((
+        match (
+            &self.args.cycle_fn,
+            &self.args.cycle_initial,
+            &self.args.cycle_result,
+        ) {
+            (Some(cycle_fn), Some(cycle_initial), None) => Ok((
                 quote!((#cycle_fn)),
                 quote!((#cycle_initial)),
                 quote!(Fixpoint),
             )),
-            (None, None) => Ok((
+            (None, None, None) => Ok((
                 quote!((salsa::plumbing::unexpected_cycle_recovery!)),
                 quote!((salsa::plumbing::unexpected_cycle_initial!)),
                 quote!(Panic),
             )),
-            (Some(_), None) => Err(syn::Error::new_spanned(
+            (Some(_), None, None) => Err(syn::Error::new_spanned(
                 self.args.cycle_fn.as_ref().unwrap(),
                 "must provide `cycle_initial` along with `cycle_fn`",
             )),
-            (None, Some(_)) => Err(syn::Error::new_spanned(
+            (None, Some(_), None) => Err(syn::Error::new_spanned(
                 self.args.cycle_initial.as_ref().unwrap(),
                 "must provide `cycle_fn` along with `cycle_initial`",
             )),
+            (None, None, Some(cycle_result)) => Ok((
+                quote!((salsa::plumbing::unexpected_cycle_recovery!)),
+                quote!((#cycle_result)),
+                quote!(FallbackImmediate),
+            )),
+            (_, _, Some(_)) => Err(syn::Error::new_spanned(
+                self.args.cycle_initial.as_ref().unwrap(),
+                "must provide either `cycle_result` or `cycle_fn` along with `cycle_initial`, not both",
```

Comment:
this error message doesn't seem right, you should not provide either `cycle_initial` or `cycle_fn` if you provide `cycle_result`

Reply:
I was also confused by this, until I realized this can be read as "provide either `cycle_result` or (`cycle_fn` along with `cycle_initial`)", but probably better to reword this.

Reply:
Oh I did not realize the ambiguity in this statement